### DISCUSSION TITLE
RunBlocking

### DISCUSSION THREAD

**Discussion by techpotatoes** [orbit-mvi/orbit-mvi#119] [General]
Hi, I've been trying to understand why is there a runBlocking {...} As part of the intent { } definition -> 

https://github.com/orbit-mvi/orbit-mvi/blob/941c3b29f8a06b4b906758d90dfc4ea49586b05f/orbit-core/src/commonMain/kotlin/org/orbitmvi/orbit/syntax/simple/SimpleSyntaxExtensions.kt

I understand it needs to wait for the response. But why not a launch on viewModelScope? Thanks in advance

**Comment by jisungbin** [comment]
I've been trying to understand this part for a long time before too, but what I got is that `container` is an extension function of `ViewModel`. It's also using `viewModelScope` inside.

```kotlin
fun <STATE : Any, SIDE_EFFECT : Any> ViewModel.container(
    initialState: STATE,
    settings: Container.Settings = Container.Settings(),
    onCreate: ((state: STATE) -> Unit)? = null
): Container<STATE, SIDE_EFFECT> {
    return viewModelScope.container(initialState, settings, onCreate)
}
```

**Reply by techpotatoes** [reply]
Hi @jisungbin , thanks for your response.

I'm not saying it's wrong or anything, I'm just trying to understand why it works like that. Because it works.. :) 

Documentation says: Calls to Container.intent` do not block the caller. The operations within are offloaded to an event-loop style background coroutine.

I think I found it, 

```
@OrbitDsl
public fun <STATE : Any, SIDE_EFFECT : Any> ContainerHost<STATE, SIDE_EFFECT>.intent(
    registerIdling: Boolean = true,
    transformer: suspend SimpleSyntax<STATE, SIDE_EFFECT>.() -> Unit
): Unit =
    runBlocking {
        container.orbit {
            withIdling(registerIdling) {
                SimpleSyntax(this).transformer()
            }
        }
    }
```

It's the "container.orbit { " bit which offloads it to another thread. 

```
/**
     * Executes an orbit intent. The intents are built in the [ContainerHost] using your chosen syntax.
     *
     * @param orbitIntent lambda returning the suspend function representing the intent
     */
    public suspend fun orbit(orbitIntent: suspend ContainerContext<STATE, SIDE_EFFECT>.() -> Unit)
 ```
 
 implemetation: 
 
 ```
     override suspend fun orbit(orbitIntent: suspend ContainerContext<STATE, SIDE_EFFECT>.() -> Unit) {
        initialiseIfNeeded()
        dispatchChannel.send(orbitIntent)
    }
 ```
 
runBlocking is going to block the main thread when you call intent, but it's a quick operation, since it's just offloading the intent to a dispatcher.  Does that make sense? 

Thanks guys

**Reply by jisungbin** [reply]
Thanks for sharing your thoughts @techpotatoes. I think I still don't understand all aspects of coroutines(I don't understand your last sentence: `but it's a quick operation, since it's just offloading the intent to a dispatcher.`). There is no end to studying coroutines!
